### PR TITLE
feat: Add pin/unpin commands to persist important buffers

### PR DESCRIPTION
**Pull request by al-ce**
Great idea for a plugin! I saw [this comment](https://www.reddit.com/r/neovim/comments/12qgs5o/comment/jgq2tlf/?utm_source=share&utm_medium=web2x&context=3) in the reddit thread and slapped a quick pin/unpin command together to address having to 'poke' important buffers with a quick switch to Insert mode. Don't know if this is something you're interested in or if it's something you'd rather implement yourself from scratch, but I thought I'd share.

The threshold is respected. Say the threshold is 4 and you have four buffers open. You pin the oldest buffer. Open a new buffer, the second oldest buffer gets popped. And so on up until you pin four buffers. Then if you open a new buffer, your four pinned ones persist, and the new one also opens. You pin the fifth and open a new one, five persist, the new one also opens, and so on. Unpinned buffers seem to be removed with respect to their actual age along others that weren't pinned in the first place.

I didn't add completion because I don't know nvim apis well enough to implement it with relatively few lines of code.

### CODE REVIEW

File: lua/hbac/init.lua
Comment (axkirillov) [inline lua/hbac/init.lua]:
```suggestion
	nargs = 1,
	complete = function()
			-- return completion candidates as a list-like table
			return {1 = "pin", 2 = "unpin", 3 = "close_unused"}
		end
```
If you want to add completion, you can do it like this

File: lua/hbac/init.lua
Comment (axkirillov) [inline lua/hbac/init.lua]:
```suggestion
	{
	 nargs = 1,
	}
)
```
Do you use a formatter? I'm using the default lsp one, I think indentation is off. I would also put the closing bracket on its own line. It's a nitpick tho 😄

File: lua/hbac/init.lua
Comment (axkirillov) [inline lua/hbac/init.lua]: On the close_unused this would print "Hbac: buffer closed_unused**ned**" 🤔

**Comment by axkirillov** [comment]
Hi! Thank you for the PR! 
Since some people do seem to prefer a mapping instead of i<esc>, I find this functionality will be useful. Furthermore the addition of unpinning a buffer is interesting, although I don't know if I would personally use it.
Lastly, since you are adding functionality, there need to be updates to the README as well.

File: lua/hbac/init.lua
Comment (al-ce) [inline lua/hbac/init.lua]:
```suggestion
	if arg == "close_unused" then
		close_unused()
		print("Hbac: Unused buffers closed")
	elseif M.commands[arg] then
		M.commands[arg]()
		print("Hbac: Buffer " .. arg .. "ned")
```

🤦  How about this? And are you ok with `Hbac` vs. `hbac` in this message? For the command itself, I was following the capitalization convention.

File: lua/hbac/init.lua
Comment (al-ce) [inline lua/hbac/init.lua]:
```suggestion
	nargs = 1,
	complete = function()
		-- return completion candidates as a list-like table
		return {[1] = "pin", [2] = "unpin", [3] = "close_unused"}
	end,
```

Your suggestion works perfectly but my LSP was complaining without the brackets. (`lua_ls` 3.6.18). Yours looks better but just thought I'd mention it. Is it for Lua 5.1 compatibility?

File: lua/hbac/init.lua
Comment (al-ce) [reply lua/hbac/init.lua:123]: Not a nitpick! I'm using `stylua` 0.17.1 and that's what it formatted for me. However you want to do it.

Comment (al-ce) [reply lua/hbac/init.lua]: Not a nitpick! I'm using `stylua` 0.17.1 and that's what it formatted for me. However you want to do it.

File: README.md
Comment (al-ce) [inline README.md]: Would you prefer `.close_unused()`, `.pin()`, and `.unpin()`, or some other names, for that lua structure?

File: lua/hbac/init.lua
Comment (axkirillov) [reply lua/hbac/init.lua]: Hbac makes sense, since all user commands in vim have to start with a capital letter

File: lua/hbac/init.lua
Comment (axkirillov) [reply lua/hbac/init.lua]: Lol, no. I'm just not that good at lua. Also is it possible to just write `return {"pin", "unpin", "close_unused"}` without the indexes?

File: lua/hbac/init.lua
Comment (axkirillov) [reply lua/hbac/init.lua]: I checked and my formatter does not care whether I use }) or
```lua
  }
)
```
I just think that the newline version looks better 😄

File: README.md
Comment (axkirillov) [reply README.md]: I like pin and unpin. It's very clear and concise. What came to mind though is that the fist command could really be called `close_unpinned`. That's more or less what it does, the difference in "used" and "pinned" is semantic, because both are handled the same. By editing a buffer you automatically "pin" it. That makes me think that also the "persist" metaphor that I used initially should be renamed to pin as well. Cause otherwise there is "pin", "use", and "persist" which all mean the same thing 🤔

File: lua/hbac/init.lua
Comment (axkirillov) [inline lua/hbac/init.lua]: Are you using some kind of AI assisted development? 😆 Otherwise the comment is superfluous 
```suggestion

```

File: lua/hbac/init.lua
Comment (axkirillov) [inline lua/hbac/init.lua]:
```suggestion
	pin = M.pin_buffer,
	unpin = M.unpin_buffer,
```
let's use the pin metaphor cause why not

File: lua/hbac/init.lua
Comment (al-ce) [reply lua/hbac/init.lua]: lol what a way to shame someone out of writing bad comments 😭 yes you're right, I used it to indicate the new section and never removed it

File: lua/hbac/init.lua
Comment (al-ce) [inline lua/hbac/init.lua]: `{"pin", "unpin", "close_unpinned"}` also work, but this way we can just pull from `commands` dynamically. Is that ok?

File: lua/hbac/init.lua
Comment (al-ce) [inline lua/hbac/init.lua]: I realize `pin` vs. `pin_buffer` may be confusing, please advise. The reason for this distinction is to only send the notification to the user when they explicitly call the `pin` command rather than during the automated pinning function.

File: lua/hbac/init.lua
Comment (al-ce) [inline lua/hbac/init.lua]: Since this is effectively a toggle of `pin_buffer`, we could refactor `pin_buffer` to also accept a second arg, maybe rename it to `toggle_pin`, and that might also resolve the confusion between `pin` vs. `pin_buffer` mentioned above. I wanted to hold off before a bigger refactor like that.

File: lua/hbac/init.lua
Comment (al-ce) [inline lua/hbac/init.lua]: With a bigger refactor redefining `M` and redefining what functions are local vs. part of `M`, this could just return `M` and only make the necessary functions available to the user. If you think this is getting too messy.

**Review by al-ce — COMMENTED** [review]
Thanks for all the feedback, and my apologies if this PR is starting to be more than you bargained for.

File: lua/hbac/init.lua
Comment (axkirillov) [reply lua/hbac/init.lua]: That's actually preferred, yeah. Bc then there is only one place where you add exported commands and you don't have to maintain autocomplete separately.

File: lua/hbac/init.lua
Comment (axkirillov) [reply lua/hbac/init.lua]: I agree, everything exported should be part of `M`, everything else can be local. That way the `M.pin` becomes the user command that triggers notification and local `pin` just pins the buffer.

File: lua/hbac/init.lua
Comment (axkirillov) [reply lua/hbac/init.lua]: Right, this could be just one function `toggle_pin`. I didn't get why would it need a second arg. It should only need the `bufnr` I think.

File: lua/hbac/init.lua
Comment (axkirillov) [reply lua/hbac/init.lua]: As mentioned in other thread `pin_bufer` should be a `local func pin`, that resolves naming conflict. Furthermore, if you agree to refactor it to `local function toggle_pin`, it can return a string message, which you can use to construct a notification or completely ignore when used internally in an autocommand

File: lua/hbac/init.lua
Comment (al-ce) [inline lua/hbac/init.lua]: This is what I was thinking when I mentioned adding a second arg. I may not have been communicating my ideas clearly or I'm missing something obvious. If you don't like this, the prior commit still sets this to a local function and has the module return a clean `M` table, and we can work from there.

File: lua/hbac/init.lua
Comment (al-ce) [inline lua/hbac/init.lua]:
```suggestion
	M.pinned_buffers[bufnr] = action == "pin"
```

`~= "unpin"` is an artifact from the buggy prior commit, will fix if you decide this refactor is acceptable

**Comment by axkirillov** [comment]
@al-ce hi, I added a bunch of refactors in the last commit, check them out please, and tell me if you find them sensible. It is mainly about modularization, but there are some improvements to the code also. On the user side I though that I can expose just the toggle command without separate pin unpin actions, tell me if that is ok for you.

File: lua/hbac/state.lua
Comment (axkirillov) [inline lua/hbac/state.lua:6]: oops, sorry 😅

File: lua/hbac/setup.lua
Comment (axkirillov) [inline lua/hbac/setup.lua:25]: Thank you for the fix, you were faster then me 😅

**Comment by al-ce** [comment]
The refactors are great, sorry I misunderstood what you were saying about exposing a single command instead of two. That makes sense. I had considered splitting up the modules but I didn't know how you preferred it. This looks good!

Thanks for working with me on this! I enjoyed the learning experience, you gave great guidance. Btw let me know if you need me to do any commit squashing on my end

File: lua/hbac/setup.lua
Comment (al-ce) [reply lua/hbac/setup.lua:25]: I figured this was probably because in the old version you had `bufnr = bufnr or vim.api.nvim_get_current_buf` in the pinning func that was making passing bufnr as an arg unnecessary in some places. I didn't add it back in so there was no need to make further changes to your refactors in the notification function. This is cleaner overall imo

File: lua/hbac/setup.lua
Comment (al-ce) [inline lua/hbac/setup.lua]: With `bufnr` instead of `buf`, the oldest buffer would close when a new buffer was open after passing the threshold, even if the oldest was pinned(manually or automatically). Is this change correct?

Comment (axkirillov) [reply lua/hbac/setup.lua]: Right, buf is correct, since we are trying to find the oldest unpinned buffer. Bufnr is just the current one.

**Comment by axkirillov** [comment]
Thank you as well, I enjoyed this too. Your commit messages are great, I am much more sloppy with them. I find squashing and rebasing a good practice, but I'm also fine with just merging all of it into main.
If you are interested in such a thing I can make you a collaborator on this repo. Otherwise I'll just merge it and create a release myself.

**Comment by al-ce** [comment]
Let's go with best practices and keep the commit history minimal and easy to read. Should I force push the rebase to this branch or can you do it all when you merge?

**Comment by axkirillov** [comment]
@al-ce Feel free to force push here and tell me when you are ready for me to merge

File: lua/hbac/setup.lua
Comment (al-ce) [inline lua/hbac/setup.lua]: I made this change and pushed it with the squash, previously it was:

```lua
command.vim_cmd_func,
```

This wasn't passing the table of subcommands to `vim_cmd_func`. Sorry I didn't notice it before, I was only using the Lua api to test since the refactor and forgot to test the Ex commands. I also renamed the `args` param in `vim_cmd_fun` to `cmd` so this function wouldn't have `args.args`. Let me know if that's acceptable.

File: lua/hbac/command.lua
Comment (al-ce) [inline lua/hbac/command.lua]: Renamed `args` param to `cmd` to be consistent and less confusing in the `nvim_create_user_command` function. Please confirm this is acceptable before merge.

**Review by al-ce — COMMENTED** [review]
Squashed but made a fix I didn't catch before, sorry for the last minute changes!

File: lua/hbac/setup.lua
Comment (axkirillov) [reply lua/hbac/setup.lua]: Nice catch. I think here `args.args` is actually correct.
From nvim documentation:

> The function is called with a single table argument that contains the following keys:
name: (string) Command name
args: (string) The args passed to the command, if any

So the first args is the `single table argument` while the second is a key in this table representing the arguments passed to the command itself

File: lua/hbac/command.lua
Comment (axkirillov) [reply lua/hbac/command.lua]: how about `name`? just throwing it out there bc the multiple versions of the word command throughout can be confusing

File: lua/hbac/setup.lua
Comment (axkirillov) [inline lua/hbac/setup.lua]: I just noticed another bug I introduced with the refactor. This has to be moved before
```lua
if not opts.autoclose then
  return
end
``` 
Otherwise the command wont be created for people who opt out of autoclose

File: lua/hbac/command.lua
Comment (al-ce) [reply lua/hbac/command.lua]: The reason i changed it to `cmd` was to be consistent with the `cmd.args` I introduced above, but since we'll do `args.args` then we can change it back to how it was in your refactor, `arg`.

**Review by axkirillov — APPROVED** [review]

**Pull request merged**
